File: src/token_saver/tools/output_pruner.py
from __future__ import annotations

import subprocess

from token_saver.filters.ansi import strip_ansi
from token_saver.filters.build_tools import detect_and_filter_build
from token_saver.filters.git import filter_git_output
from token_saver.filters.test_runners import detect_and_filter_tests
from token_saver.utils.token_counter import estimate_tokens

MAX_STREAM_BYTES = 2 * 1024 * 1024  # 2MB runaway output buffer ceiling
STREAM_HEAD_BYTES = 1024 * 1024  # Keep first 1MB
STREAM_TAIL_BYTES = 512 * 1024  # Keep last 512KB


def filter_generic(output: str) -> str:
    return output


def auto_filter(output: str, exit_code: int = 0) -> str:
    test_filtered = detect_and_filter_tests(output)
    if test_filtered is not None:
        return test_filtered

    build_filtered = detect_and_filter_build(output)
    if build_filtered is not None:
        return build_filtered

    if "git " in output[:100] or "commit" in output or "branch" in output:
        return filter_git_output(output)

    return filter_generic(output)


def filter_output_logic(raw_output: str, output_type: str = "auto", exit_code: int = 0) -> str:
    raw_output = raw_output or ""
    # Memory ceiling guard: protect host process against runaway infinite output streams
    if len(raw_output) > MAX_STREAM_BYTES:
        truncated_count = len(raw_output) - (STREAM_HEAD_BYTES + STREAM_TAIL_BYTES)
        raw_output = (
            raw_output[:STREAM_HEAD_BYTES]
            + f"\n\n... [Token-Saver Stream Guard: Truncated {truncated_count:,} bytes of runaway output to protect memory] ...\n\n"
            + raw_output[-STREAM_TAIL_BYTES:]
        )

    clean_output = strip_ansi(raw_output)

    if output_type == "auto":
        filtered = auto_filter(clean_output, exit_code)
    elif output_type == "pytest":
        from token_saver.filters.test_runners import filter_pytest

        filtered = filter_pytest(clean_output)
    elif output_type == "jest":
        from token_saver.filters.test_runners import filter_jest_vitest

        filtered = filter_jest_vitest(clean_output)
    elif output_type == "npm":
        from token_saver.filters.build_tools import filter_npm_yarn

        filtered = filter_npm_yarn(clean_output)
    elif output_type == "cargo":
        from token_saver.filters.build_tools import filter_cargo

        filtered = filter_cargo(clean_output)
    elif output_type == "git":
        filtered = filter_git_output(clean_output)
    elif output_type == "generic":
        filtered = filter_generic(clean_output)
    else:
        filtered = clean_output

    # Fallback Safety Guard:
    # If the command failed (exit_code != 0), guarantee critical error context is never lost.
    if exit_code != 0:
        error_keywords = (
            "traceback",
            "error",
            "failed",
            "exception",
            "fatal",
            "panic",
            "cannot",
            "syntaxerror",
            "importerror",
        )
        raw_has_error = any(kw in clean_output.lower() for kw in error_keywords)
        filtered_has_error = any(kw in filtered.lower() for kw in error_keywords)

        if (raw_has_error and not filtered_has_error) or not filtered.strip():
            filtered = clean_output.strip() + "\n[Token-Saver: Preserved full error context due to non-zero exit code]"

    if not filtered.strip() and clean_output.strip():
        filtered = clean_output.strip()

    orig_tokens = estimate_tokens(raw_output)
    new_tokens = estimate_tokens(filtered)
    pct = 0
    if orig_tokens > 0:
        pct = int((orig_tokens - new_tokens) / orig_tokens * 100)

    footer = f"\n[Token-Saver: {orig_tokens} -> {new_tokens} tokens ({pct}% saved)]"
    if orig_tokens > new_tokens:
        try:
            from token_saver.telemetry.stats import tracker

            tracker.record_savings("command", orig_tokens, new_tokens)
        except Exception:
            pass
    return filtered + footer


def register_output_pruner_tools(mcp):
    @mcp.tool()
    def run_command_smart(
        command: str,
        cwd: str = ".",
        timeout: int = 120,
        background: bool = False,
    ) -> str:
        """Executes a shell command and returns intelligently filtered output.
        Use this tool instead of raw shell commands when you want to minimize token usage
        from verbose CLI outputs like tests, builds, and package managers.

        Set background=True to launch dev servers, daemons, or long-running watchers
        without blocking the agent.
        """
        if background:
            try:
                proc = subprocess.Popen(
                    command,
                    cwd=cwd,
                    shell=True,
                    stdout=subprocess.DEVNULL,
                    stderr=subprocess.DEVNULL,
                )
                return f"[BACKGROUND PROCESS LAUNCHED] PID: {proc.pid} | Command: {command}"
            except Exception as e:
                return f"Error launching background command: {e}"

        try:
            result = subprocess.run(
                command,
                cwd=cwd,
                shell=True,
                capture_output=True,
                timeout=timeout,
            )
            stdout_str = (result.stdout or b"").decode("utf-8", errors="replace")
            stderr_str = (result.stderr or b"").decode("utf-8", errors="replace")
            raw_output = f"{stdout_str}\n{stderr_str}".strip() if stderr_str else stdout_str
            exit_code = result.returncode
            filtered = filter_output_logic(raw_output, output_type="auto", exit_code=exit_code)
            return f"Exit Code: {exit_code}\n" + filtered
        except subprocess.TimeoutExpired as e:
            stdout_text = (
                e.stdout
                if isinstance(e.stdout, str)
                else (e.stdout.decode("utf-8", errors="replace") if e.stdout else "")
            )
            stderr_text = (
                e.stderr
                if isinstance(e.stderr, str)
                else (e.stderr.decode("utf-8", errors="replace") if e.stderr else "")
            )
            raw_output = f"{stdout_text}\n{stderr_text}".strip()
            filtered = filter_output_logic(raw_output, output_type="auto", exit_code=-1)
            return f"Command timed out after {timeout}s\n" + filtered
        except Exception as e:
            return f"Error executing command: {str(e)}"

    @mcp.tool()
    def filter_output(raw_output: str, output_type: str = "auto") -> str:
        """Filters raw text output to save tokens.
        output_type can be: "auto", "pytest", "jest", "npm", "cargo", "git", "generic".
        Useful when you already have output from another source.
        """
        return filter_output_logic(raw_output, output_type)

File: crates/tokenjar-core/src/output_pruner.rs
//! Output Pruner & Command Runner Engine in Rust.
//!
//! Intelligently filters command line outputs (pytest, cargo test, npm, jest)
//! and terminal logs to prevent token explosion.

use std::process::{Command, Stdio};

use crate::filters::ansi::strip_ansi;
use crate::filters::build_tools::{detect_and_filter_build, filter_npm_yarn};
use crate::filters::git::filter_git_output;
use crate::filters::test_runners::{filter_cargo, filter_pytest};
use crate::telemetry::TelemetryTracker;
use crate::token_counter::estimate_tokens;

const MAX_STREAM_BYTES: usize = 2 * 1024 * 1024; // 2MB
const STREAM_HEAD_BYTES: usize = 1024 * 1024; // 1MB
const STREAM_TAIL_BYTES: usize = 512 * 1024; // 512KB

/// Core output filtering logic.
pub fn filter_output_logic(
    raw_output: &str,
    output_type: &str,
    exit_code: i32,
    tracker: &TelemetryTracker,
) -> String {
    let mut clean_raw = raw_output.to_string();

    // Memory ceiling guard
    if clean_raw.len() > MAX_STREAM_BYTES {
        let mut head_idx = STREAM_HEAD_BYTES.min(clean_raw.len());
        while head_idx > 0 && !clean_raw.is_char_boundary(head_idx) {
            head_idx -= 1;
        }
        let mut tail_idx = clean_raw.len().saturating_sub(STREAM_TAIL_BYTES);
        while tail_idx < clean_raw.len() && !clean_raw.is_char_boundary(tail_idx) {
            tail_idx += 1;
        }
        if head_idx < tail_idx {
            let truncated_count = tail_idx - head_idx;
            let head = &clean_raw[..head_idx];
            let tail = &clean_raw[tail_idx..];
            clean_raw = format!(
                "{head}\n\n... [TokenJar Stream Guard: Truncated {truncated_count} bytes of runaway output to protect memory] ...\n\n{tail}"
            );
        }
    }

    let clean = strip_ansi(&clean_raw);

    let mut filtered = match output_type {
        "pytest" => filter_pytest(&clean),
        "cargo" => filter_cargo(&clean),
        "npm" | "yarn" | "pnpm" => filter_npm_yarn(&clean),
        "git" => filter_git_output(&clean),
        "generic" => clean.clone(),
        _ => {
            // Auto-detect
            if clean.contains("pytest")
                || clean.contains("=== test session starts ===")
                || clean.contains("passed in")
            {
                filter_pytest(&clean)
            } else if clean.contains("running ") && clean.contains("test result:") {
                filter_cargo(&clean)
            } else if let Some(b) = detect_and_filter_build(&clean) {
                b
            } else if clean.contains("git ")
                || clean.contains("On branch ")
                || clean.contains("Untracked files:")
            {
                filter_git_output(&clean)
            } else {
                clean.clone()
            }
        }
    };

    // Fallback safety guard: preserve full error context if exit_code != 0
    if exit_code != 0 {
        let error_keywords = [
            "traceback",
            "error",
            "failed",
            "exception",
            "fatal",
            "panic",
            "cannot",
            "syntaxerror",
            "importerror",
        ];
        let raw_lower = clean.to_lowercase();
        let filtered_lower = filtered.to_lowercase();
        let raw_has_error = error_keywords.iter().any(|&k| raw_lower.contains(k));
        let filtered_has_error = error_keywords.iter().any(|&k| filtered_lower.contains(k));

        if (raw_has_error && !filtered_has_error) || filtered.trim().is_empty() {
            filtered = format!(
                "{}\n[TokenJar: Preserved full error context due to non-zero exit code]",
                clean.trim()
            );
        }
    }

    if filtered.trim().is_empty() && !clean.trim().is_empty() {
        filtered = clean.trim().to_string();
    }

    let orig_tokens = estimate_tokens(&clean_raw);
    let new_tokens = estimate_tokens(&filtered);
    let pct = if orig_tokens > 0 {
        ((orig_tokens.saturating_sub(new_tokens) as f64 / orig_tokens as f64) * 100.0) as usize
    } else {
        0
    };

    if orig_tokens > new_tokens {
        tracker.record_savings("command", orig_tokens as u64, new_tokens as u64);
    }

    format!("{filtered}\n[TokenJar: {orig_tokens} -> {new_tokens} tokens ({pct}% saved)]")
}

/// Runs a command via the system shell with intelligent token filtering and timeout protection.
pub fn run_command_smart(
    command_str: &str,
    cwd: &str,
    timeout_secs: u64,
    background: bool,
    tracker: &TelemetryTracker,
) -> String {
    if background {
        #[cfg(target_os = "windows")]
        let mut cmd = Command::new("cmd");
        #[cfg(target_os = "windows")]
        cmd.args(["/C", command_str]);

        #[cfg(not(target_os = "windows"))]
        let mut cmd = Command::new("sh");
        #[cfg(not(target_os = "windows"))]
        cmd.args(["-c", command_str]);

        cmd.current_dir(cwd)
            .stdout(Stdio::null())
            .stderr(Stdio::null());

        match cmd.spawn() {
            Ok(child) => format!(
                "[BACKGROUND PROCESS LAUNCHED] PID: {} | Command: {command_str}",
                child.id()
            ),
            Err(e) => format!("Error launching background command: {e}"),
        }
    } else {
        #[cfg(target_os = "windows")]
        let mut cmd = Command::new("cmd");
        #[cfg(target_os = "windows")]
        cmd.args(["/C", command_str]);

        #[cfg(not(target_os = "windows"))]
        let mut cmd = Command::new("sh");
        #[cfg(not(target_os = "windows"))]
        cmd.args(["-c", command_str]);

        cmd.current_dir(cwd)
            .stdout(Stdio::piped())
            .stderr(Stdio::piped());

        let mut child = match cmd.spawn() {
            Ok(c) => c,
            Err(e) => return format!("Error executing command: {e}"),
        };

        let stdout_handle = child.stdout.take();
        let stderr_handle = child.stderr.take();

        let stdout_thread = std::thread::spawn(move || {
            let mut buf = Vec::new();
            if let Some(mut h) = stdout_handle {
                use std::io::Read;
                let _ = h.read_to_end(&mut buf);
            }
            buf
        });

        let stderr_thread = std::thread::spawn(move || {
            let mut buf = Vec::new();
            if let Some(mut h) = stderr_handle {
                use std::io::Read;
                let _ = h.read_to_end(&mut buf);
            }
            buf
        });

        let start = std::time::Instant::now();
        let timeout = std::time::Duration::from_secs(timeout_secs.max(1));
        let exit_code;

        loop {
            match child.try_wait() {
                Ok(Some(status)) => {
                    exit_code = status.code().unwrap_or(-1);
                    break;
                }
                Ok(None) => {
                    if start.elapsed() >= timeout {
                        let _ = child.kill();
                        let _ = child.wait();
                        let _ = stdout_thread.join();
                        let _ = stderr_thread.join();
                        return format!("Error: Command timed out after {timeout_secs} seconds.");
                    }
                    std::thread::sleep(std::time::Duration::from_millis(20));
                }
                Err(e) => return format!("Error waiting for command: {e}"),
            }
        }

        let stdout_bytes = stdout_thread.join().unwrap_or_default();
        let stderr_bytes = stderr_thread.join().unwrap_or_default();
        let stdout_str = String::from_utf8_lossy(&stdout_bytes).to_string();
        let stderr_str = String::from_utf8_lossy(&stderr_bytes).to_string();
        let combined = if stderr_str.is_empty() {
            stdout_str
        } else if stdout_str.is_empty() {
            stderr_str
        } else {
            format!("{stdout_str}\n{stderr_str}")
        };

        let filtered = filter_output_logic(&combined, "auto", exit_code, tracker);
        format!("Exit Code: {exit_code}\n{filtered}")
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_filter_output_logic_cargo() {
        let raw = r#"
   Compiling tokenjar v0.8.0
    Finished test [unoptimized + debuginfo] target(s) in 0.12s
     Running unittests src/lib.rs

running 3 tests
test test_a ... ok
test test_b ... ok
test test_c ... ok

test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
"#;
        let temp_telemetry = tempfile::NamedTempFile::new().unwrap();
        let tracker = TelemetryTracker::with_path(temp_telemetry.path().to_path_buf());
        let result = filter_output_logic(raw, "cargo", 0, &tracker);
        assert!(result.contains("test result: ok"));
        assert!(result.contains("TokenJar:"));
    }

    #[test]
    fn test_command_timeout() {
        let temp_telemetry = tempfile::NamedTempFile::new().unwrap();
        let tracker = TelemetryTracker::with_path(temp_telemetry.path().to_path_buf());
        #[cfg(target_os = "windows")]
        let cmd = "powershell -NoProfile -Command Start-Sleep -Seconds 4";
        #[cfg(not(target_os = "windows"))]
        let cmd = "sleep 4";

        let result = run_command_smart(cmd, ".", 1, false, &tracker);
        assert!(result.contains("timed out"));
    }
}
